crates/ruff/src/rules/pyflakes/rules/f_string_missing_placeholders.rs

```diff
@@ -47,45 +47,72 @@ impl AlwaysAutofixableViolation for FStringMissingPlaceholders {
-fn find_useless_f_strings<'a>(
-    expr: &'a Expr,
+/// Return an iterator containing a two-element tuple for each f-string in the
+/// given [`ExprFString`] expression.
+///
+/// The first element of the tuple is the f-string prefix range, and the second
+/// element is the entire f-string range. It returns an iterator because of the
+/// possibility of multiple f-strings implicitly concatenated together.
+///
+/// For example,
+///
+/// ```python
+///   f"first" rf"second"
+/// # ^         ^            (prefix range)
+/// # ^^^^^^^^ ^^^^^^^^^^    (token range)
+/// ```
+///
+/// would return `[(0..1, 0..8), (10..11, 9..19)]`.
+///
+/// This function assumes that the given f-string expression is without any
+/// placeholder expressions.
+///
+/// [`ExprFString`]: `ruff_python_ast::ExprFString`
+fn fstring_prefix_and_tok_range<'a>(
+    fstring: &'a ast::ExprFString,
     locator: &'a Locator,
     source_type: PySourceType,
 ) -> impl Iterator<Item = (TextRange, TextRange)> + 'a {
-    let contents = locator.slice(expr);
-    lexer::lex_starts_at(contents, source_type.as_mode(), expr.start())
+    let contents = locator.slice(fstring);
+    let mut current_f_string_start = fstring.start();
+    lexer::lex_starts_at(contents, source_type.as_mode(), fstring.start())
         .flatten()
-        .filter_map(|(tok, range)| match tok {
-            Tok::String {
-                kind: StringKind::FString | StringKind::RawFString,
-                ..
-            } => {
-                let first_char = locator.slice(TextRange::at(range.start(), TextSize::from(1)));
+        .filter_map(move |(tok, range)| match tok {
+            Tok::FStringStart => {
+                current_f_string_start = range.start();
+                None
+            }
+            Tok::FStringEnd => {
+                let first_char =
+                    locator.slice(TextRange::at(current_f_string_start, TextSize::from(1)));
                 // f"..."  => f_position = 0
                 // fr"..." => f_position = 0
                 // rf"..." => f_position = 1
                 let f_position = u32::from(!(first_char == "f" || first_char == "F"));
                 Some((
                     TextRange::at(
-                        range.start() + TextSize::from(f_position),
+                        current_f_string_start + TextSize::from(f_position),
                         TextSize::from(1),
                     ),
-                    range,
+                    TextRange::new(current_f_string_start, range.end()),
                 ))
             }
             _ => None,
         })
 }
 
 /// F541
-pub(crate) fn f_string_missing_placeholders(expr: &Expr, values: &[Expr], checker: &mut Checker) {
+pub(crate) fn f_string_missing_placeholders(
+    fstring: &ast::ExprFString,
+    values: &[Expr],
+    checker: &mut Checker,
+) {
     if !values
         .iter()
         .any(|value| matches!(value, Expr::FormattedValue(_)))
```

Comment:
```suggestion
    if values.is_empty()
```

Because values can only be initialized with formatted values?

Reply:
No, it's initialized with both formatted values and string literals (other than the expression part). So, `f"foo"` will give:
```rust
String(
    ExprFString {
        range: 0..6,
        values: [
            Constant(
                ExprConstant {
                    range: 2..5,
                    value: Str(
                        StringConstant {
                            value: "foo",
                            unicode: false,
                            implicit_concatenated: false,
                        },
                    ),
                },
            ),
        ],
        implicit_concatenated: false,
    },
)
```